Add section heads to help commands.

richsalz opened the pull request:
Fixes: #9936 

This is a WIP.  When done I'll remove the TODO file.

richsalz commented:
If someone can take a look at the output, I'd like an opinion of there should be blank line before each section.

richsalz commented:
First pass done; all programs with >=20 help lines: asn1parse ca cms crl dgst ecparam enc ocsp pkcs12 pkcs8 pkeyutl req rsa rsautl s_client s_server smime speed ts verify x509

richsalz commented:
All commands done, out of WIP, ready for review.

richsalz commented:
in order to entice folks to look at and review this, here's some output:
```
; ./apps/openssl rsa --help
Usage: rsa [options]

General options:
 -help              Display this summary
 -check             Verify key consistency
 -*                 Any supported cipher
 -engine val        Use engine, possibly a hardware device

Input options:
 -in val            Input file
 -inform format     Input format, one of DER PEM
 -pubin             Expect a public key in input file
 -RSAPublicKey_in   Input is an RSAPublicKey
 -passin val        Input file pass phrase source

Output options:
 -out outfile       Output file
 -outform format    Output format, one of DER PEM PVK
 -pubout            Output a public key
 -RSAPublicKey_out  Output is an RSAPublicKey
 -passout val       Output file pass phrase source
 -noout             Don't print key out
 -text              Print the key in text
 -modulus           Print the RSA key modulus

PVK options:
 -pvk-strong        Enable 'Strong' PVK encoding level (default)
 -pvk-weak          Enable 'Weak' PVK encoding level
 -pvk-none          Don't enforce PVK encoding
; 
; ./apps/openssl rand --help
Usage: rand [flags] num

General options:
 -help               Display this summary
 -engine val         Use engine, possibly a hardware device

Output options:
 -out outfile        Output file
 -base64             Base64 encode output
 -hex                Hex encode output

Random state options:
 -rand val           Load the file(s) into the random number generator
 -writerand outfile  Write random data to the specified file
```

For common options (OPT_[VSXR]_OPTIONS in ```apps/lib/opt.c```), as shown in the "Random state options" text above, the prolog is common text.

t8m commented on apps/fipsinstall.c:
```diff
@@ -32,18 +32,26 @@ typedef enum OPTION_choice {
+    OPT_SECTION("General"),
     {"help", OPT_HELP, '-', "Display this summary"},
+#if 0
+    /* Gross ugly long line that belongs in the manpage. */
```
Just delete it and add it to the manpage unless the example is already there.

richsalz replied:
the manpage has a couple of complete examples so i removed this.

t8m commented on apps/pkcs8.c:
```diff
@@ -31,30 +31,37 @@ typedef enum OPTION_choice {
 
 const OPTIONS pkcs8_options[] = {
     {"help", OPT_HELP, '-', "Display this summary"},
```
Missing General options section marker.

richsalz replied:
good catch!

richsalz commented:
fixup commit pushed.

levitte commented on apps/ciphers.c:
```diff
@@ -31,31 +31,34 @@ typedef enum OPTION_choice {
     {"s", OPT_S, '-', "Only supported ciphers"},
 #ifndef OPENSSL_NO_SSL3
-    {"ssl3", OPT_SSL3, '-', "SSL3 mode"},
+    {"ssl3", OPT_SSL3, '-', "Ciphers defined for SSL3"},
```
Is this, and the similar ones following, really output options per se?  They rather seem like selectors

richsalz replied:
Good point; I changed it to "Cipher selection"

levitte commented on apps/asn1pars.c:
```diff
@@ -27,27 +27,32 @@ typedef enum OPTION_choice {
     {"help", OPT_HELP, '-', "Display this summary"},
+    {"oid", OPT_OID, '<', "file of extra oid definitions"},
+
+    OPT_SECTION("I/O"),
```
In others, you have separated Input from Output.  Not here, is there a reason?

richsalz replied:
see https://github.com/openssl/openssl/pull/9953#issuecomment-539992422

levitte commented on apps/ca.c:
```diff
@@ -154,10 +154,30 @@ typedef enum OPTION_choice {
+    {"in", OPT_IN, '<', "The input PEM encoded cert request(s)"},
+    {"infiles", OPT_INFILES, '-', "The last argument, requests to process"},
+    {"out", OPT_OUT, '>', "Where to put the output file(s)"},
+    {"notext", OPT_NOTEXT, '-', "Do not print the generated certificate"},
```
For other commands, the input and output options have their own sections, but not this one:

richsalz replied:
see https://github.com/openssl/openssl/pull/9953#issuecomment-539992422

richsalz commented on apps/asn1pars.c:
```diff
@@ -27,27 +27,32 @@ typedef enum OPTION_choice {
+
+    OPT_SECTION("Formatting"),
+    {"i", OPT_INDENT, 0, "indents the output"},
+    {"noout", OPT_NOOUT, 0, "do not produce any output"},
```
fixed.

richsalz commented:
As for the Input, Output, Input/Output: if there were a bunch of options, or I thought it made sense, I divided them into sections, otherwise I didn't.  I agree it might be better to go for consistency, but I also dislike a section that has only one option (yes sometimes that is unavoidable).

I hope that this topic can be addressed in a separate PR.  (Probably by someone else :)

t8m commented on apps/ciphers.c:
```diff
@@ -34,7 +34,7 @@ const OPTIONS ciphers_options[] = {
     {"help", OPT_HELP, '-', "Display this summary"},
 
-    OPT_SECTION("Output"),
+    OPT_SECTION("Cipher specification"),
```
Hmm, should there be two separate groups? Output (-v, -V, -stdname, -convert) and Cipher selection (for the rest)?

richsalz replied:
yes, there should :)  additional commit pushed.  thanks.

t8m commented:
My approval still holds.

t8m commented:
ping @openssl/omc for 2nd review

richsalz commented:
Another sample, to entice reviewers:
```
; ./apps/openssl pkcs7 --help
Usage: pkcs7 [options]

General options:
 -help             Display this summary
 -engine val       Use engine, possibly a hardware device

Input options:
 -in infile        Input file
 -inform PEM|DER   Input format - DER or PEM

Output options:
 -outform PEM|DER  Output format - DER or PEM
 -out outfile      Output file
 -noout            Don't output encoded data
 -text             Print full details of certificates
 -print            Print out all fields of the PKCS7 structure
 -print_certs      Print_certs  print any certs or crl in the input
; 
```

levitte commented:
Travis failures seem relevant, there seems to be some option duplication in some commands.

t8m commented:
Yeah, the rebase broke the apps/enc.c. There are added duplicate options in, out, and pass.

t8m commented:
OK, re-approved now.
ping @levitte ?

richsalz commented:
Another sample:
```
; ./apps/openssl ocsp --help
Usage: ocsp [options]

General options:
 -help                   Display this summary
 -ignore_err             Ignore error on OCSP request or response and continue running
 -CAfile infile          Trusted certificates file
 -CApath infile          Trusted certificates directory
 -no-CAfile              Do not load the default certificates file
 -no-CApath              Do not load certificates from the default certificates directory

Responder options:
 -timeout +int           Connection timeout (in seconds) to the OCSP responder
 -resp_no_certs          Don't include any certificates in response
 -multi +int             run multiple responder processes
 -no_certs               Don't include any certificates in signed request
 -badsig                 Corrupt last byte of loaded OSCP response signature (for test)
 -CA infile              CA certificate
 -nmin +int              Number of minutes before next update
 -nrequest +int          Number of requests to accept (default unlimited)
 -reqin val              File with the DER-encoded request
 -signer infile          Certificate to sign OCSP request with
 -sign_other infile      Additional certificates to include in signed request
 -index infile           Certificate status index file
 -ndays +int             Number of days before next update
 -rsigner infile         Responder certificate to sign responses with
 -rkey infile            Responder key to sign responses with
 -rother infile          Other certificates to include in response
 -rmd val                Digest Algorithm to use in signature of OCSP response
 -rsigopt val            OCSP response signature parameter in n:v form
 -header val             key=value header to add
 -rcid val               Use specified algorithm for cert id in response
 -*                      Any supported digest algorithm (sha1,sha256, ... )

Client options:
 -url val                Responder URL
 -host val               TCP/IP hostname:port to connect to
 -port +int              Port to run responder on
 -out outfile            Output filename
 -noverify               Don't verify response at all
 -nonce                  Add OCSP nonce to request
 -no_nonce               Don't add OCSP nonce to request
 -no_signature_verify    Don't check signature on response
 -resp_key_id            Identify response by signing certificate key ID
 -no_cert_verify         Don't check signing certificate
 -text                   Print text form of request and response
 -req_text               Print text form of request
 -resp_text              Print text form of response
 -no_chain               Don't chain verify response
 -no_cert_checks         Don't do additional checks on signing certificate
 -no_explicit            Do not explicitly check the chain, just verify the root
 -trust_other            Don't verify additional certificates
 -no_intern              Don't search certificates contained in response for signer
 -respin val             File with the DER-encoded response
 -VAfile infile          Validator certificates file
 -verify_other infile    Additional certificates to search for signer
 -path val               Path to use in OCSP request
 -cert infile            Certificate to check
 -serial val             Serial number to check
 -validity_period ulong  Maximum validity discrepancy in seconds
 -signkey val            Private key to sign OCSP request with
 -reqout val             Output file for the DER-encoded request
 -respout val            Output file for the DER-encoded response
 -issuer infile          Issuer certificate
 -status_age +int        Maximum status age in seconds

Validation options:
 -policy val             adds policy to the acceptable policy set
 -purpose val            certificate chain purpose
 -verify_name val        verification policy name
 -verify_depth int       chain depth limit
 -auth_level int         chain authentication security level
 -attime intmax          verification epoch time
 -verify_hostname val    expected peer hostname
 -verify_email val       expected peer email
 -verify_ip val          expected peer IP address
 -ignore_critical        permit unhandled critical extensions
 -issuer_checks          (deprecated)
 -crl_check              check leaf certificate revocation
 -crl_check_all          check full chain revocation
 -policy_check           perform rfc5280 policy checks
 -explicit_policy        set policy variable require-explicit-policy
 -inhibit_any            set policy variable inhibit-any-policy
 -inhibit_map            set policy variable inhibit-policy-mapping
 -x509_strict            disable certificate compatibility work-arounds
 -extended_crl           enable extended CRL features
 -use_deltas             use delta CRLs
 -policy_print           print policy processing diagnostics
 -check_ss_sig           check root CA self-signatures
 -trusted_first          search trust store first (default)
 -suiteB_128_only        Suite B 128-bit-only mode
 -suiteB_128             Suite B 128-bit mode allowing 192-bit algorithms
 -suiteB_192             Suite B 192-bit-only mode
 -partial_chain          accept chains anchored by intermediate trust-store CAs
 -no_alt_chains          (deprecated)
 -no_check_time          ignore certificate validity time
 -allow_proxy_certs      allow the use of proxy certificates
; 
```

t8m commented:
For the record my approval still holds.

richsalz commented:
I had to rebase this because #8442 added new flags to several commands.  There were, of course, merge conflicts.  No other changes were made.

Seeking reviews.  First opened September 19.  No feedback (except poor patient @t8m :) for nearly a month.

t8m commented:
My approval still holds. It would be nice if @openssl/omc provided the second review soon to avoid further merge conflicts.

richsalz commented on apps/ciphers.c:
```diff
@@ -31,33 +31,38 @@ typedef enum OPTION_choice {
 #endif
 #ifndef OPENSSL_NO_TLS1_2
-    {"tls1_2", OPT_TLS1_2, '-', "TLS1.2 mode"},
+    {"tls1_2", OPT_TLS1_2, '-', "Ciphers defined for TLS1.2"},
```
done.

richsalz commented on apps/ciphers.c:
```diff
@@ -31,33 +31,38 @@ typedef enum OPTION_choice {
 #endif
 #ifndef OPENSSL_NO_TLS1_3
-    {"tls1_3", OPT_TLS1_3, '-', "TLS1.3 mode"},
+    {"tls1_3", OPT_TLS1_3, '-', "Ciphers defined for TLS1.3"},
```
done.

richsalz commented on apps/ciphers.c:
```diff
@@ -31,33 +31,38 @@ typedef enum OPTION_choice {
 #endif
 #ifndef OPENSSL_NO_TLS1
-    {"tls1", OPT_TLS1, '-', "TLS1 mode"},
+    {"tls1", OPT_TLS1, '-', "Ciphers defined for TLS1"},
```
done.

richsalz commented on apps/ocsp.c:
```diff
@@ -147,31 +147,69 @@ typedef enum OPTION_choice {
-    {"noverify", OPT_NOVERIFY, '-', "Don't verify response at all"},
-    {"nonce", OPT_NONCE, '-', "Add OCSP nonce to request"},
-    {"no_nonce", OPT_NO_NONCE, '-', "Don't add OCSP nonce to request"},
+    {"CAfile", OPT_CAFILE, '<', "Trusted certificates file"},
```
I did each command without thinking too much of overall consistency.  I agree that more shuffling around can and should be done.  I would rather see that done as a followup PR.

vdukhovni replied:
Is that PR forthcoming? I am somewhat reluctant to have this merged if the followup PR may not materialize.

richsalz replied:
While I am likely to look at this, I make no commitment.  I think this PR, *as it currently stands* is a big improvement over what OpenSSL currently has.  I will shrug my shoulders (and scratch my head in wonderment :) if you don't think so, too.

richsalz commented on apps/req.c:
```diff
@@ -140,18 +126,41 @@ const OPTIONS req_options[] = {
+    {"pkeyopt", OPT_PKEYOPT, 's', "Public key options as opt:value"},
+    {"sigopt", OPT_SIGOPT, 's', "Signature parameter in n:v form"},
+    {"verify", OPT_VERIFY, '-', "Verify signature on REQ"},
+    {"modulus", OPT_MODULUS, '-', "RSA modulus"},
```
fixed thanks.

richsalz commented on apps/req.c:
```diff
@@ -140,18 +126,41 @@ const OPTIONS req_options[] = {
+    {"newkey", OPT_NEWKEY, 's', "Specify as type:bits"},
+    {"pkeyopt", OPT_PKEYOPT, 's', "Public key options as opt:value"},
+    {"sigopt", OPT_SIGOPT, 's', "Signature parameter in n:v form"},
+    {"verify", OPT_VERIFY, '-', "Verify signature on REQ"},
```
I moved it to the "general" section.  Not great, admittedly, but better.

I also moved in/inform there, and out/outform  to the "output" section.

richsalz commented on apps/s_client.c:
```diff
@@ -664,51 +691,37 @@ const OPTIONS s_client_options[] = {
-    {"maxfraglen", OPT_MAXFRAGLEN, 'p',
-     "Enable Maximum Fragment Length Negotiation (len values: 512, 1024, 2048 and 4096)"},
-    {"fallback_scsv", OPT_FALLBACKSCSV, '-', "Send the fallback SCSV"},
     {"name", OPT_PROTOHOST, 's',
```
yes, that's better.  moved.

richsalz commented on apps/s_time.c:
```diff
@@ -92,6 +79,25 @@ const OPTIONS s_time_options[] = {
+    {"nameopt", OPT_NAMEOPT, 's', "Various certificate name options"},
+    {"cert", OPT_CERT, '<', "Cert file to use, PEM format assumed"},
+    {"key", OPT_KEY, '<', "File with key, PEM; default is -cert file"},
+    {"cafile", OPT_CAFILE, '<', "PEM format file of CA's"},
```
Same; I am not doing this now.

richsalz commented on apps/smime.c:
```diff
@@ -52,39 +52,62 @@ const OPTIONS smime_options[] = {
     {"text", OPT_TEXT, '-', "Include or delete text MIME headers"},
+    {"nosmimecap", OPT_NOSMIMECAP, '-', "Omit the SMIMECapabilities attribute"},
+
+    OPT_SECTION("Certificate chain"),
```
Same reply, I am not doing this for this PR.

richsalz commented on apps/x509.c:
```diff
@@ -65,37 +65,60 @@ typedef enum OPTION_choice {
+     "Print old-style (MD5) subject hash value"},
+#endif
+
+    OPT_SECTION("Certificate"),
```
The only output-related option I could see is "-nameopt" which I moved to output.  Did you8 have others in mind?

richsalz commented on apps/x509.c:
```diff
@@ -104,51 +127,35 @@ const OPTIONS x509_options[] = {
+    {"force_pubkey", OPT_FORCE_PUBKEY, '<', "Force the key to put inside certificate"},
+    {"subj", OPT_SUBJ, 's', "Set or override certificate subject (and issuer)"},
+
+    OPT_SECTION("CA"),
```
Can you give me more details?  Which options should move, and to which section?

richsalz commented on apps/verify.c:
```diff
@@ -36,10 +36,17 @@ typedef enum OPTION_choice {
         "Print extra information about the operations being performed."},
+    {"nameopt", OPT_NAMEOPT, 's', "Various certificate name options"},
+
+    OPT_SECTION("Certificate chain"),
```
Same reply, I am not doing this for this PR.

t8m commented:
I believe we should focus on real errors/mistakes and not try to achieve perfection in this PR. IMO  even in the state as it is now it would be major improvement over the current situation.

richsalz commented:
I don't understand your question.

richsalz commented:
(PS:  Want to tweak the labels so the 24hour hold timer works?)

paulidale commented:
I'm wondering why there are OPT_MORE_STR lines in the text.  They seem to be manually putting line breaks in places -- this could be automated.

It's quite poissible I missing something egregious.

richsalz commented:
The number really didn't change.  Auto-wrapping the text would be more trouble than it's worth for the roughly-a-dozen lines:
```
; g m
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
; g grep OPT_MORE_STR | wc 
     14     105     859
; g co app-help-categories 
Switched to branch 'app-help-categories'
; g grep OPT_MORE_STR | wc 
     13      91     750
; 
```

paulidale commented:
Merged to master.  Thanks for the effort and endurance on of this one.
